grpc/src/main/java/io/stargate/grpc/service/QueryHandler.java

```diff
@@ -0,0 +1,258 @@
+import io.stargate.proto.QueryOuterClass.QueryParameters;
+import io.stargate.proto.QueryOuterClass.Response;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import org.apache.cassandra.stargate.db.ConsistencyLevel;
+import org.apache.cassandra.stargate.exceptions.PreparedQueryNotFoundException;
+
+class QueryHandler extends MessageHandler {
+
+  private final Query query;
+  private final PrepareInfo prepareInfo;
+
+  QueryHandler(
+      Query query,
+      Connection connection,
+      Cache<PrepareInfo, CompletionStage<Prepared>> preparedCache,
+      Persistence persistence,
+      StreamObserver<Response> responseObserver) {
+    super(connection, preparedCache, persistence, responseObserver);
+    this.query = query;
+    QueryParameters queryParameters = query.getParameters();
+    this.prepareInfo =
+        ImmutablePrepareInfo.builder()
+            .keyspace(
+                queryParameters.hasKeyspace() ? queryParameters.getKeyspace().getValue() : null)
+            .user(connection.loggedUser().map(AuthenticatedUser::name).orElse(null))
+            .cql(query.getCql())
+            .build();
+  }
+
+  void handle() {
+    CompletionStage<Result> resultFuture = prepare(prepareInfo).thenCompose(this::executePrepared);
+
+    // If our local cache got out of sync with the server, we might get an UNPREPARED response,
+    // reprepare on the fly and retry:
+    resultFuture = handleUnprepared(resultFuture, this::reprepareAndRetry);
+
+    resultFuture
+        .thenApply(this::buildResponse)
+        .thenCompose(this::executeTracingQueryIfNeeded)
+        .whenComplete(
+            (response, error) -> {
+              if (error != null) {
+                handleException(error);
+              } else {
+                setSuccess(response);
+              }
+            });
+  }
+
+  private CompletionStage<Result> executePrepared(Prepared prepared) {
+    long queryStartNanoTime = System.nanoTime();
+
+    Payload values = query.getValues();
+    PayloadHandler handler = PayloadHandlers.get(values.getType());
+
+    QueryParameters parameters = query.getParameters();
+
+    try {
+      return connection.execute(
+          bindValues(handler, prepared, values), makeParameters(parameters), queryStartNanoTime);
+    } catch (Exception e) {
+      CompletableFuture<Result> failedFuture = new CompletableFuture<>();
+      failedFuture.completeExceptionally(e);
+      return failedFuture;
+    }
+  }
+
+  private CompletionStage<Result> reprepareAndRetry() {
+    return prepare(prepareInfo, true).thenCompose(this::executePrepared);
+  }
+
+  private ResponseAndTraceId buildResponse(Result result) {
+    ResponseAndTraceId responseAndTraceId = new ResponseAndTraceId();
+    responseAndTraceId.setTracingId(result.getTracingId());
```

Comment:
The original code also checks that the tracing flag is set in `query.getParameters()`. I think we can reasonably assume that the response contains an id if and only if the client asked for it.